system/include/emscripten/em_asm.h

```diff
@@ -1,13 +1,72 @@
 #ifndef __em_asm_h__
 #define __em_asm_h__
 
-// The wasm backend calls vararg functions by passing the variadic arguments as
-// an array on the stack. For EM_ASM's underlying functions, s2wasm needs to
-// translate them to emscripten_asm_const_[signature], but the expected
-// signature and arguments has been lost as part of the vararg buffer.
-// Therefore, we declare EM_ASM's implementing functions as non-variadic.
-
 #ifndef __asmjs
+// When calling EM_ASM functions, we're calling out to JS. JS only has doubles
+// natively, so we can convert all C types to double before the call. By
+// converting all arguments to double, we can store them in a vararg buffer,
+// and we only need to pass the number of args in order to read them in JS.
+
+// When converting arguments to double, we need to:
+//   - upcast floats to doubles, so we have consistent bit widths
+//   - convert ints to doubles (via f64.convert_s/i32 and such)
+//   - cast pointers to ints, and then convert (C disallows direct pointer to
+//     double conversions)
+// We can use the _Generic clang extension / C11 feature to decide at compile
```

Comment:
Let's use the more-correct (and more importantly, more-searchable) term "generic selection" here. And maybe calling it a C11 feature first (which as an extension clang supports everywhere) makes it sound less dodgy :)